ref(contributors): update contributor roles and requirements (#4776)
openservicemesh/osm: CONTRIBUTOR_LADDER.md

Each of the roles is organized into the following categories:
* "Responsibilities" list what the role is expected to do
* "Requirements" are qualifications the role needs to meet
* "Privileges" list what the role is entitled to

| Role | Responsibilities | Requirements | Privileges |
| -----| ---------------- | ------------ | -------|
| [Maintainer](#maintainer) | Active member of the community |<ul><li>Reviewer for ≥3 months</li><li> Demonstrates deep understanding of multiple areas of the project</li><li>Exercise good judgement for the project</li><li>Maintain contributions to the project</li></ul> | <ul><li>Has GitHub or CI/CD rights to approve any pull requests</li><li>Represent the project as a Maintainer</li><li>Communicate with CNCF on behalf of the project</li></ul> |

> | [Reviewer](#reviewer) | Review contributions from other members | <ul><li>Contributor for ≥3 months</li><li>Demonstrates knowledge in an area of the project</li><li>Reviews pull requests</li></ul> | Has GitHub or CI/CD rights to approve pull requests in specific directories |

Reviewer: @shashankram I agree with that. The goal right now is to incentivize reviews and I think reducing the complexity of the change (not adding another role) and streamlining the process to become a Maintainer will help with that.

@shalier what are your thoughts on this? I think we should describe the process of adding maintainers as codeowners for certain subdirectories and define the process of how to gain more ownership. We'll also need to update the maintainer role to reflect these changes. 

I think we can revisit adding more contributor roles as the OSM project grows.

> | [Reviewer](#reviewer) | Review contributions from other members | <ul><li>Contributor for ≥3 months</li><li>Demonstrates knowledge in an area of the project</li><li>Reviews pull requests</li></ul> | Has GitHub or CI/CD rights to approve pull requests in specific directories |

Reviewer: @shalier yeah, let's hold off

> | [Reviewer](#reviewer) | Review contributions from other members | <ul><li>Contributor for ≥3 months</li><li>Demonstrates knowledge in an area of the project</li><li>Reviews pull requests</li></ul> | Has GitHub or CI/CD rights to approve pull requests in specific directories |

Reviewer 2: Hi, I would like to comment on a new role, currently reviewing PRs takes more time as fewer maintainers are there for the project. By adding a reviewer role, it will be easier for a new contributor to come up to the maintainer role as the path will be more clear and more transparent. Also, it will help in faster reviewing.

> | [Reviewer](#reviewer) | Review contributions from other members | <ul><li>Contributor for ≥3 months</li><li>Demonstrates knowledge in an area of the project</li><li>Reviews pull requests</li></ul> | Has GitHub or CI/CD rights to approve pull requests in specific directories |

Reviewer: @mudit-01 I think the proposed changes to the maintainer role will make it easier to become a maintainer since ownership doesn't need to be over the entire project. A contributor can be promoted to a maintainer when they demonstrate the requirements of a maintainer in at least one project area. Hopefully, the review process will be faster with more maintainers!

Each of the roles is organized into the following categories:
* "Responsibilities" list what the role is expected to do
* "Requirements" are qualifications the role needs to meet
* "Privileges" list what the role is entitled to

| Role | Responsibilities | Requirements | Privileges |
| -----| ---------------- | ------------ | -------|
| [Maintainer](#maintainer) | Active member of the community |<ul><li>Contributor for ≥3 months</li><li> Demonstrates deep understanding of multiple areas of the project</li><li>Exercise good judgement for the project</li><li>Maintain contributions to the project</li></ul> | <ul><li>Has GitHub or CI/CD rights to approve any pull requests</li><li>Represent the project as a Maintainer</li><li>Communicate with CNCF on behalf of the project</li></ul> |
| [Contributor](#contributor) | Following the [CNCF Code of Conduct] and the project [contributing guide] | One or more contributions to the project | Invitations to contributor events |

> | [Community Participant](#community-participant) | Following the [CNCF Code of Conduct] | --- | --- |

Reviewer: What do you think about reversing this list so the order matches the order in which the roles are presented below? It will also clarify the role progression of community participant to maintainer.

    * Promotes the project in public
* Privileges:
    * Invitations to contributor events
    * Eligible to become a Maintainer

Process of becoming a Contributor:
1. There is no nomination process for this role. Being a contributor entails being a community member who does one or more of the listed requirements.

## Maintainer

> Description: Maintainers are established contributors who are responsible for the entire project. As such, they have the ability to approve PRs against any area of the project, and are expected to participate in making decisions about the strategy and priorities of the project. Maintainers focus on a holistic review of contributions: performance, compatibility, adherence to convention, and overall quality.

Reviewer: ```suggestion
Description: Maintainers are established contributors who are responsible for one or more project areas. They have the ability to approve PRs against the project areas they own and are expected to participate in making decisions about the strategy and priorities of the project.
```

    * Determining strategy and policy for the project
    * Participating in, and leading, community meetings
    * Helps run the project infrastructure
* Requirements:
    * Experience as a Contributor for at least 3 months
    * Deep understanding of the technical goals and direction of the project

>     * Deep understanding of the technical domain of the project across multiple areas

Reviewer: ```suggestion
    * Deep understanding of the technical domain of the project across their areas of ownership
```

    * Sustained contributions to design and direction by doing all of:
      * Authoring and reviewing proposals (GitHub Issues for refactoring, enhancements, or new functionality)
      * Initiating, contributing, and resolving discussions (emails, GitHub issues, meetings)
      * Identifying subtle or complex issues in designs and implementation PRs
    * Is able to exercise judgement for the good of the project, independent of their employer, social circles, or teams
    * Mentors other Contributors
* Additional privileges:

>     * Approve PRs to any area of the project

Reviewer: ```suggestion
    * Approve PRs to areas of ownership
```

    * Represent the project in public as a Maintainer
    * Communicate with the CNCF on behalf of the project
    * Have a vote in Maintainer decisions

Process of becoming a Maintainer:

> 1. A current contributor may be self-nominated or be nominated by a current Maintainer by opening a PR against the root of the [OSM repository] and adding the nominee to the [OWNERS file] under the *owners* entry. Additionally, if the nomination is for a code owner whose PR approvals are meant to satisfy the PR merge requirements, the nominee must be added to the [CODEOWNERS file] in the same PR. A codeowner is defined per the [GitHub definition](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#about-code-owners).

Reviewer: In this PR, I think we can clarify a maintainer's ability to be an owner of a sub-project area rather than the entire project.

```suggestion
1. A current contributor may be self-nominated or be nominated by a current Maintainer by opening a PR against the root of the [OSM repository] and adding the nominee to the [OWNERS file] under the *owners* entry. Additionally, if the nomination is for a code owner whose PR approvals are meant to satisfy the PR merge requirements against specific areas of the codebase, the nominee must be added to the [CODEOWNERS file] for their specific areas of ownership in the same PR. A codeowner is defined per the [GitHub definition](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#about-code-owners).
```

2. The nominee will add a comment to the PR testifying that they agree to all requirements of becoming a Maintainer.
3. A majority of the current Maintainers must then approve the PR.

> 

Reviewer: We could also define the process of how to add areas of ownership.

Author: Added an example here: https://github.com/openservicemesh/osm/pull/4776/files#diff-653bc99dc2e48c5310ac61677b097abbaa6193149342b09bc721e058e3aae91bR91
